src/linker/Linker/LinkerILProcessor.cs

```diff
@@ -0,0 +1,125 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using Mono.Cecil;
+using Mono.Cecil.Cil;
+
+namespace Mono.Linker
+{
+	class LinkerILProcessor
+	{
+		readonly ILProcessor _ilProcessor;
+
+		Collections.Generic.Collection<Instruction> Instructions => _ilProcessor.Body.Instructions;
+
+		internal LinkerILProcessor (MethodBody body)
+		{
+			_ilProcessor = body.GetILProcessor ();
+		}
+
+		public void Emit (OpCode opcode) => Append (_ilProcessor.Create (opcode));
+		public void Emit (OpCode opcode, TypeReference type) => Append (_ilProcessor.Create (opcode, type));
+		public void Emit (OpCode opcode, MethodReference method) => Append (_ilProcessor.Create (opcode, method));
+		public void Emit (OpCode opcode, VariableDefinition variable) => Append (_ilProcessor.Create (opcode, variable));
+
+		public void Emit (OpCode opcode, string value) => Append (_ilProcessor.Create (opcode, value));
+		public void InsertBefore (Instruction target, Instruction instruction)
+		{
+			// So when inserting before, all pointers have to be updated to the new "before" instruction.
+			RedirectScopeStart (target, instruction);
+			_ilProcessor.InsertBefore (target, instruction);
+		}
+
+		// When inserting after, no redirection is necessary since it will naturally be "appended" to the same blocks
+		// as the target instruction.
+		public void InsertAfter (Instruction target, Instruction instruction)
+		{
+			RedirectScopeEnd (target, instruction);
+			_ilProcessor.InsertAfter (target, instruction);
+		}
+
+		public void Append (Instruction instruction)
+		{
+			Instruction lastInstruction = Instructions.Count == 0 ? null : Instructions[Instructions.Count - 1];
+			RedirectScopeEnd (lastInstruction, instruction);
+			_ilProcessor.Append (instruction);
+		}
+
+		public void Replace (Instruction target, Instruction instruction)
+		{
+			RedirectScopeStart (target, instruction);
+			RedirectScopeEnd (target, instruction);
+			ReplaceInstruction (target, instruction);
+			_ilProcessor.Replace (target, instruction);
+		}
+
+		public void Replace (int index, Instruction instruction) => Replace (_ilProcessor.Body.Instructions[index], instruction);
+
+		public void Remove (Instruction instruction)
+		{
+			int index = _ilProcessor.Body.Instructions.IndexOf (instruction);
+			if (index == -1)
+				throw new ArgumentOutOfRangeException (nameof (instruction));
+
+			Instruction nextInstruction = Instructions.Count == index + 1 ? null : Instructions[index + 1];
+			Instruction previousInstruction = index == 0 ? (Instructions.Count == 1 ? null : Instructions[index + 1]) : Instructions[index - 1];
+
+			RedirectScopeStart (instruction, nextInstruction);
+			RedirectScopeEnd (instruction, previousInstruction);
```

Comment:
Should removing an instruction also fix up branch targets?

Reply:
Yes - it should - good catch!